functions/src/modules/email.functions.ts

```diff
@@ -0,0 +1,100 @@
+
+const path = require('path');
+const os = require('os');
+const fs = require('fs');
+
+
+let emailTemplates = {};
+
+
+
+/////////////////////////////////////////////////
+//          Update Email Templates on sendInBlue 
+//          in response to 
+//          1. changes of the templates on FB (Storage Triggered)
+//          2. changes of its metadata ( DB Update Trigger )
+//
+/////////////////////////////////////////////////
+
+
+const updateEmailTemplates = async (filePath) => {
+    // slice(0, -9) to get rid of the trailing '.mustache' extension
+    const fileName = filePath.split('/')[2].slice(0, -9);
+    const tempLocalFile = path.join(os.tmpdir(), fileName);
+    
+    
+    if (!filePath.startsWith('email/templates')) 
+        return;
+
+    let emailRef = bucket.file(filePath);
+    
+    await emailRef.download({ destination: tempLocalFile });
+    
+    // Use the following to upate the template on SendInBlue
+    let htmlContent = fs.readFileSync(tempLocalFile, 'utf-8');
+
+    
+    // 1. Get the template ID (sendInBlue ID)
+    let templateNode = await db.ref('/email/templates').orderByKey()
+                                .equalTo(fileName).once('value');
+    let template = templateNode.val();
+
+    if (!template.exists()) {
+        console.log("Template metadata doesn't exist.");
+        return;
+    }
+    
+    let { sender, subject } = template;
+
+    let id = helpers.getTemplateId(fileName);
+    
+    if (+id === -1) { // New Template
+        id = await helpers.createTemplate(fileName, sender, htmlContent, subject);
+        await db.ref(`/email/templates/${fileName}`)
+                .update({ lastUpdated: new Date() });
+    }            
+    else {
+        await helpers.updateTemplate(id, htmlContent);
+        await db.ref(`/email/templates/${template.key}`)
+                .update({ lastUpdated: new Date() });
+    }
+
+    // 2. Send a test Email confirming it has been updated correctly
+    helpers.sendTestTemplate(id, sender.email);        
+}
+
+export const updateTemplatesOnTemplateUpload = functions.storage.object()
+.onFinalize(object => { // called when either a NEW object is created, or when an object is overwritten
+    updateEmailTemplates(object.name);
+    return 1;
+});
+
+export const updateTemplatesOnMetadataChange = functions.database.ref('/email/templates}')
+.onUpdate(async (change, context) => {    
+    updateEmailTemplates(change.after.key);
+    return 1;
+});
+
+
+
+export const sendNotificationEmail = functions.database.ref('/email/notifications}')
+.onCreate(async (snapshot, context) => {
+    const data = snapshot.val();
+    const templateName = snapshot.key;
+
+    let id;
+    if (Object.keys(emailTemplates).indexOf(templateName) > -1)
+        id = emailTemplates[templateName].id;
+    else 
+        id = await helpers.getTemplateId(templateName)
+
+    await helpers.sendEmail(data.to, data.bcc, id, data.params);
+
+    return snapshot.ref.update({ sent: true });
```

Comment:
I think it might be more useful to save the timestamp when the notification was actually sent. It would help debugging in the future. Of course, it should be named differently, like `sentTimestamp`.

And please test that this property is not filled to prevent double sending.

Reply:
I'm not actually getting the point out of using a flag of any kind here... the Firebase function watches only **newly created** nodes .. I created it as you suggested now .. but I'm just asking

Reply:
Maybe I'm overthinking. The flag might be needed in case we need to add a notification that should not be sent… As in importing from some old system, as we done for SQR, do you remember? Just in case, I want to be absolutely sure that the notification would not be sent twice.